runtime.md

```diff
@@ -34,21 +34,19 @@ See [Query State](#query-state) for information on retrieving the state of a con
 ## Lifecycle
 The lifecycle describes the timeline of events that happen from when a container is created to when it ceases to exist.
 
-1. OCI compliant runtime is invoked with a reference to the location of the bundle.
-   How this reference is passed to the runtime is an implementation detail.
-2. The container's runtime environment MUST be created according to the configuration in [`config.json`](config.md).
-   Any updates to `config.json` after container is running MUST not affect the container.
-3. The prestart hooks MUST be invoked by the runtime.
-   If any prestart hook fails, then the container MUST be stopped and the lifecycle continues at step 8.
-4. The user specified process MUST be executed in the container.
-5. The poststart hooks MUST be invoked by the runtime.
-   If any poststart hook fails, then the container MUST be stopped and the lifecycle continues at step 8.
-6. Additional actions such as pausing the container, resuming the container or signaling the container MAY be performed using the runtime interface.
+1. OCI compliant runtime's `create` command is invoked with a reference to the location of the bundle and a unique identifier.
+   How these references are passed to the runtime is an implementation detail.
+2. The container's runtime environment (namespaces, mounts, etc.) MUST be created according to the configuration in [`config.json`](config.md).
+   If a new PID namespace is requested in the [`config.json`](config.md), a PID namespace MUST created at this time.
+   However, the user-specified code (from ['process'](config.md#process-configuration) MUST NOT be executed at this time.
+   With the exception of the user-specified process, any updates to `config.json` after container is created MUST NOT affect the container.
+3. Runtime's `start` command is invoked with the unique identifier of the container.
+   The runtime MUST create and start the user-specified code, as specified by [`process`](config.md#process-configuration), in the container's PID namespace.
+   Any updates to the user-specified code in ['process'](config.md#process-configuration) after this point MUST NOT have any effect on the container.
+4. Additional actions such as pausing the container, resuming the container or signaling the container MAY be performed using the runtime interface.
    The container MAY also error out, exit or crash.
-7. The container MUST be destroyed by undoing the steps performed during create phase (step 2).
-8. The poststop hooks MUST be invoked by the runtime and errors, if any, MAY be logged.
-
-Note: The lifecycle is a WIP and it will evolve as we have more use cases and more information on the viability of a separate create phase.
+5. Irrespective of how the user defined process stops (i.e. PID 1 exits), once the PID namespace is deleted the container MUST be destroyed by undoing the steps performed during create phase (step 2).
+   For clarity, all namespaces that were created in step 2 MUST be deleted.
```

Comment:
I'd rather leave this clarification line off.  For example, the caller may have bind-mounted one or more of those namespaces somewhere else.  In cases like that, I think the bind-mounter has assumed responsibility for cleaning up the pinned resource.  The runtime is only responsible for removing any pinning that it has setup.

The approach in the previous paragraph works for cases where the kernel is handling refcounting and cleanup (when the last consumer is removed).  It's trickier for things like cgroups, which aren't  automatically removed when the last consumer goes away.  In that case, I'm fine with any of:

a. The runtime makes a best-effort attempt to cleanup cgroups, logging warnings (not errors) if it can't delete a cgroup it created (because some other processes are still using that cgroups).
b. The runtime MUST delete any cgroups it created, killing (or moving to another cgroup) any remaining processes as it sees fit.

(a) is more in line with my preferred namespace handling (the agent that added the other processes assumes responsibility for the cleanup).  But (b) is fine with me too (mostly because I don't care that much about runtime-managed cgroups ;).